moby/moby
# Make one-shot stats faster

**- What I did**

Kuberenets CRI interface `ListContainerStats` requires to collect every pod container's stats. Currently `daemon/stats/collector.go` runs in a serial way and sleep `1s` at the end of `for` loop. This causes `ListContainerStats` to suffer long delays if there are many pod containers on the node. 

This PR is to address this issue.

**- How I did it**

- Make the stats collector parallel with the number of CPUs.
- ~`Sleep` is not needed in a normal loop since a condition variable is used. Only sleep when an error happens at system level, then retry in next loop~. `time.Sleep(s.interval)` is needed for `one-short=0` or `stream=1`, otherwise the caller would receive successive stats without any interval.
- Move `bufReader` out of `Collector`, since it cannot be shared by multiple goroutines.
- Get container stats directly (i.e. without using publishers) for `one-shot` requests. 

**- How to verify it**

Run the PR build with `cri-dockerd`, observe the time taken for `ListContainerStats`. Without this code change, it could take over 3 seconds when about 40 pod containers on the node. ~With this fix, it takes less than 1 second~. 
- Making stats collection parallel reduces the `ListContainerStats` time to 2 seconds on the 4-CPU node. 
- Getting container stats directly for `one-shot` requests reduces the `ListContainerStats` time to less than 200ms. (The requests sent by `cri-dockerd` are `one-shot` requests)  

**- Description for the changelog**

Improves the performance of the stats collector

**- A picture of a cute animal (not mandatory but encouraged)**
